tests/test_network.py

```diff
@@ -124,6 +125,106 @@ def test_add_objective(factory, client):
     assert dataset.objective_key == objective.key
 
 
+@pytest.mark.parametrize('asset_name', [
+    'dataset',
+    'objective',
+    'algo',
+    'aggregate_algo',
+    'composite_algo',
+])
+@pytest.mark.parametrize('metadata,metadata_output', [
+    ({'foo': 'bar'}, {'foo': 'bar'}),
+    (None, {}),
+    ({}, {}),
+])
+def test_asset_with_metadata(factory, client, asset_name, metadata, metadata_output):
+    create_spec = getattr(factory, f"create_{asset_name}")
+    add_asset = getattr(client, f"add_{asset_name}")
+
+    spec = create_spec(metadata=metadata)
+    asset = add_asset(spec)
+
+    assert asset.metadata == metadata_output
+
+
+@pytest.mark.parametrize('asset_name', [
+    'dataset',
+    'objective',
+    'algo',
+    'aggregate_algo',
+    'composite_algo',
+])
+@pytest.mark.parametrize('metadata', [
+    {'foo' * 40: "bar"},
+    {"foo": 'bar' * 40},
+])
+def test_asset_with_invalid_metadata(factory, client, asset_name, metadata):
+    create_spec = getattr(factory, f"create_{asset_name}")
+    add_asset = getattr(client, f"add_{asset_name}")
+
+    spec = create_spec(metadata=metadata)
+
+    with pytest.raises(substra.exceptions.InvalidRequest):
+        add_asset(spec)
+
+
+@pytest.mark.parametrize('asset_name,algo_type', [
+    ('traintuple', 'algo'),
+    ('composite_traintuple', 'composite_algo'),
+])
+@pytest.mark.parametrize('metadata,metadata_output', [
+    ({'foo': 'bar'}, {'foo': 'bar'}),
+    (None, {}),
+    ({}, {}),
+])
+def test_traintuple_with_metadata(factory, client, asset_name, algo_type, default_dataset, metadata, metadata_output):
```

Comment:
As said in a previous comment I think we should remove this test as well as the next one (`test_traintuple_with_invalid_metadata`).

In my opinion, the end to end tests should not test all the use cases and all the corner cases. This should be done in the unit test. So for maintainability and performance reasons, I think they are not needed.